### PR TITLE
fix(jit-html): add convention for html attributes

### PR DESCRIPTION
**Pull request by bigopon**
# Pull Request

## 📖 Description

Basic work to add back convention for `.bind`

### 🎫 Issues

N/A

## 👩‍💻 Reviewer Notes

N/A

## 📑 Test Plan

N/A

## ⏭ Next Steps

N/A

### CODE REVIEW

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://25290-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://25303-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by fkleuver** [comment]
I think this is pretty good as it is. Maybe rename to `IAttrSyntaxTransformer`?

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://25422-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://25437-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://25443-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by EisenbergEffect** [comment]
Yeah! Another step closer to making thing feel like Aurelia. With this and the PR from @3cp we'll pretty much be there. I took a quick look and it seems fine to me. Nice refactoring of the attr mappings as well. @fkleuver may have more specific feedback if he wants things adjusted. If he gives it the thumbs up, then let's get this in.

File: packages/jit-html/src/attribute-syntax-transformer.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+export interface IAttrSyntaxTransformer {
+  transform(node: HTMLElement, attrSyntax: AttrSyntax): void;
+  map(tagName: string, attr: string): string;
+}
+
+export const IAttrSyntaxTransformer =
+  DI
+    .createInterface<IAttrSyntaxTransformer>('IAttrSyntaxTransformer')
+    .withDefault(x => x.singleton(HtmlAttrSyntaxTransformer));
```
Comment (fkleuver) [inline packages/jit-html/src/attribute-syntax-transformer.ts]: Sorry I missed this earlier. I'd prefer to use `noDefault`. See [this comment](https://github.com/aurelia/aurelia/pull/584#discussion_r317352659) for reasoning

File: packages/jit-html/src/attribute-syntax-transformer.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            target = 'scrollLeft';
+            break;
+          case 'readonly':
+            target = 'readOnly';
+            break;
+        }
+        break;
+    }
+    return target;
+  }
```
Comment (fkleuver) [inline packages/jit-html/src/attribute-syntax-transformer.ts]: I think this would be a bit cleaner (and slightly improve perf):

```ts
  public map(tagName: string, attr: string): string {
    switch (tagName) {
      case 'LABEL':
        switch (attr) {
          case 'for':
            return 'htmlFor';
          default:
            return attr;
        }
      case 'IMG':
        switch (attr) {
          case 'usemap':
            return 'useMap';
          default:
            return attr;
        }
      case 'INPUT':
        switch (attr) {
          case 'maxlength':
            return 'maxLength';
          case 'minlength':
            return 'minLength';
          case 'formaction':
            return 'formAction';
          case 'formenctype':
            return 'formEncType';
          case 'formmethod':
            return 'formMethod';
          case 'formnovalidate':
            return 'formNoValidate';
          case 'formtarget':
            return 'formTarget';
          default:
            return attr;
        }
      case 'TEXTAREA':
        switch (attr) {
          case 'maxlength':
            return 'maxLength';
          default:
            return attr;
        }
      case 'TD':
      case 'TH':
        switch (attr) {
          case 'rowspan':
            return 'rowSpan';
          case 'colspan':
            return 'colSpan';
          default:
            return attr;
        }
      default:
        switch (attr) {
          case 'accesskey':
            return 'accessKey';
          case 'contenteditable':
            return 'contentEditable';
          case 'tabindex':
            return 'tabIndex';
          case 'textcontent':
            return 'textContent';
          case 'innerhtml':
            return 'innerHTML';
          case 'scrolltop':
            return 'scrollTop';
          case 'scrollleft':
            return 'scrollLeft';
          case 'readonly':
            return 'readOnly';
          default:
            return attr;
        }
    }
  }
```

File: packages/jit-html/src/attribute-syntax-transformer.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+      (tagName === 'TEXTAREA' || tagName === 'SELECT') && target === 'value'
+      || (target === 'textcontent' || target === 'innerhtml') && element.hasAttribute('contenteditable')
+      || target === 'scrolltop'
+      || target === 'scrollleft'
+    ) {
+      command = 'two-way';
+    }
+
+    attrSyntax.command = command;
+  }
```
Comment (fkleuver) [inline packages/jit-html/src/attribute-syntax-transformer.ts]: What about:

```ts
  function shouldDefaultToTwoWay(element: HTMLElement, attrSyntax: AttrSyntax): boolean {
    if (attrSyntax.command !== 'bind') {
      return false;
    }

    switch (element.tagName) {
      case 'INPUT':
        switch ((element as HTMLInputElement).type) {
          case 'checkbox':
          case 'radio':
            return attrSyntax.target === 'checked';
          default:
            return attrSyntax.target === 'value' || attrSyntax.target === 'files';
        }
      case 'TEXTAREA':
      case 'SELECT':
        return attrSyntax.target === 'value';
      default:
        switch (attrSyntax.target) {
          case 'textcontent':
          case 'innerhtml':
            return element.hasAttribute('contenteditable');
          case 'scrolltop':
          case 'scrolleft':
            return true;
          default:
            return false;
        }
    }
```

And then:

```ts
    if (shouldDefaultToTwoWay(node, attrSyntax)) {
      attrSyntax.command = 'two-way';
    }
```

Feel free to disagree, it's just an idea that I feel makes it a bit more readable

File: packages/jit-html/src/template-compiler.ts [hunk 4]
```diff
@@ -69,11 +70,17 @@ const { enter, leave } = Profiler.createTimer('TemplateCompiler');
  * @internal
  */
 export class TemplateCompiler implements ITemplateCompiler {
-  public static readonly inject: readonly Key[] = [ITemplateElementFactory, IAttributeParser, IExpressionParser];
+  public static readonly inject: readonly Key[] = [
+    ITemplateElementFactory,
+    IAttributeParser,
+    IExpressionParser,
+    IAttrSyntaxTransformer
+  ];
```
Comment (fkleuver) [inline packages/jit-html/src/template-compiler.ts]: Let's use parameter decorators in the ctor instead? Or maybe better to do that in bulk in a separate PR? Your call. Just felt I should mention since you reformatted already

File: packages/jit-html/src/attribute-syntax-transformer.ts (hunk 1 above)
Comment (bigopon) [reply packages/jit-html/src/attribute-syntax-transformer.ts]: I do like it better, and i think we talked about this too. Though im not sure how to organise it, separate packge or just reference the impl in index file?

File: packages/jit-html/src/attribute-syntax-transformer.ts (hunk 2 above)
Comment (bigopon) [reply packages/jit-html/src/attribute-syntax-transformer.ts]: early return should be better than value change + break

Comment (bigopon) [reply packages/jit-html/src/attribute-syntax-transformer.ts]: copy pasta 😬

File: packages/jit-html/src/attribute-syntax-transformer.ts (hunk 3 above)
Comment (bigopon) [reply packages/jit-html/src/attribute-syntax-transformer.ts]: While I was doing this, I was trying to find a beautiful name for this command transformation, hence the name. The suggestion above looks nice 👍
Though it looks it's worse for minification, but for that, probably we can do it in bulk later.

File: packages/jit-html/src/template-compiler.ts (hunk 4 above)
Comment (bigopon) [reply packages/jit-html/src/template-compiler.ts]: yeah, I'll move them to constructor

File: packages/jit-html/src/attribute-syntax-transformer.ts (hunk 1 above)
Comment (fkleuver) [reply packages/jit-html/src/attribute-syntax-transformer.ts]: Export the class itself as an `IRegistry`, and add it to `DefaultComponents` array.
Just look at `configuration.ts`.

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://25547-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://25559-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://25572-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html
